### PR TITLE
WIP: Command line interface for users to leave their reviews

### PR DESCRIPTION
**Pull request by richardbatty**
I'd like to get some feedback on this, @lpil and @jonnyarnold 

It works, but seems hard to test because of the user interaction.

### CODE REVIEW

**Comment by reevoo-samuel** [comment]
_If you're new, read the [guidelines](https://github.com/reevoo/guidelines/blob/master/pull_requests.md)._

Good code is **tested**, **designed for change**, **easy to understand** and **able to cope when
things go wrong**.

Good pull requests are **descriptive**, **small** and **short-lived**.

**For the Author**
- [x] Description explains what you have done and why.
- [ ] Tests have been written (or you have explained why you don't need
  tests).
- [ ] Documentation/comments have been added/updated.

**For the Reviewer**
- [ ] The code is well-designed and able to cope with change.
- [ ] The code is easy to understand.
- [ ] Tests cover happy and unhappy paths in the code.
- [ ] Tests are green on CI (or pulled/tested on your machine).

File: lib/cli.ex [hunk 1]
```diff
@@ -0,0 +1,40 @@
+defmodule Openreevoo.Cli do
+  @moduledoc """
+  The command line interface that users leave their review with
+  """
+
+  def run do
+    choose_library |> get_review |> format_review |> IO.puts
+  end
+
+  defp choose_library do
```
Comment (lpil) [inline lib/cli.ex:12]: It would be nicer if we could decouple the logic from the IO here, and test them separately.

File: lib/cli.ex [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    IO.puts Openreevoo.Dependencies.formatted
+    library_index_str = IO.gets "Which library would you like to review? (choose a number): "
+    {library_index, _} = library_index_str |> Integer.parse
+    Enum.at(Openreevoo.Dependencies.list, library_index)
+  end
+
+  defp get_review(library) do
+    %{
```
Comment (lpil) [inline lib/cli.ex]: Prefer a struct over a map when you know what the data is going to look like. This is worth doing in Ruby too.

File: lib/cli.ex [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    IO.gets "Out of 10, how would you rate #{library}?\n> "
+  end
+
+  defp good_points(library) do
+    IO.gets "What was good about #{library}?\n> "
+  end
+
+  defp bad_points(library) do
+    IO.gets "What was bad about #{library}?\n> "
+  end
```
Comment (lpil) [inline lib/cli.ex:34]: Is there any reason why we need these three functions? They all do the same thing. Could just iterate through a list of prompts.

**Comment by lpil** [comment]
You can test `IO.gets` by using an input device other than the process group leader. :)

https://github.com/elixir-lang/elixir/blob/v1.0.5/lib/elixir/test/elixir/io_test.exs#L53-L58

**Comment by lpil** [comment]
Writing reviews into the `IO.gets` prompt seems really uncomfortable and error prone. Can't open a temp file in `$EDITOR`, or write a template for them to fill in, and then have them submit it with the editor?

Or better yet, make a review app.

File: lib/dependencies.ex [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+  def list do
+    Mix.Dep.Lock.read
+    |> Enum.map(fn {k,_} -> to_string(k) end)
+  end
+
+  defp format(deps) do
+    deps
+    |> Enum.with_index
+    |> Enum.map(fn {library, index} -> "#{to_string(index)}: #{library}\n" end)
```
Comment (lpil) [inline lib/dependencies.ex]: Interpolation calls `to_string/1` for you.

File: lib/cli.ex [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  def run do
+    choose_library |> get_review |> format_review |> IO.puts
+  end
+
+  defp choose_library do
+    IO.puts Openreevoo.Dependencies.formatted
+    library_index_str = IO.gets "Which library would you like to review? (choose a number): "
+    {library_index, _} = library_index_str |> Integer.parse
+    Enum.at(Openreevoo.Dependencies.list, library_index)
+  end
```
Comment (lpil) [inline lib/cli.ex:18]: What happens if I type `kladughslgk` into this prompt?

Comment (jonnyarnold) [inline lib/cli.ex:18]: It breaks. Like the tests say it should ;)

Comment (lpil) [inline lib/cli.ex:18]: Ah. I see.

What _should_ happen if I type `kladughslgk` into this prompt? :D

File: lib/cli.ex (hunk 3 above)
Comment (jonnyarnold) [inline lib/cli.ex:34]: I think these functions _are_ useful - they show meaning. I'd do:

``` elixir
defp rating(library) do
  "Out of 10, how would you rate #{library}?" <> prompt
end

defp good_points(library) do
  "What was good about #{library}?" <> prompt
end

defp bad_points(library) do
  "What was bad about #{library}?" <> prompt
end

defp prompt do
  "\n> "
end
```

**Comment by jonnyarnold** [comment]
Most people don't have `$EDITOR` set up (checked with me and @richardbatty), and this is the simplest thing we can do to try out the idea.

**Pull request merged**